File: py-algorithm/number/find_the_pivot_integer.py
import unittest
from typing import List
from queue import PriorityQueue
from collections import Counter

"""
    https://leetcode.com/problems/find-the-pivot-integer/
    
    Số pivot là số chia đoạn [1,n] thành 2 đoạn thỏa mãn sum([1,x]) = sum([x,n])
    Cho số n, tìm số pivot. Nếu không có trả về -1.
"""


class Solution(unittest.TestCase):

    @staticmethod
    def pivotInteger(n: int) -> int:
        total = n * (n + 1) // 2
        left_sum = 0
        for i in range(n, 0, -1):
            left_sum += i
            right_sum = total - left_sum + i
            if left_sum == right_sum: return i

        return -1


    def test(self):
        self.assertEqual(6, self.pivotInteger(8))
        self.assertEqual(-1, self.pivotInteger(4))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/java/number/FindThePivotInteger.java
package number;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/*
    https://leetcode.com/problems/find-the-pivot-integer/

    Số pivot là số chia đoạn [1,n] thành 2 đoạn sao cho sum([1,x]) = sum([x,n])

    Cho số n, tìm số pivot. Nếu không có thì trả về -1.
 */
public class FindThePivotInteger {

    /**
     * Trick ở đây là x thường nằm cuối nên duyệt từ cuối về sẽ nhanh hơn đáng kể.
     */
    public int pivotInteger(int n) {
        int sum = n * (n + 1) / 2;
        int leftSum = 0;
        for (int i = n; i >= 1; i--) {
            leftSum += i;
            int rightSum = sum - leftSum + i;
            if (leftSum == rightSum) return i;
        }
        return -1;
    }

    @Test
    public void test() {
        assertEquals(6, pivotInteger(8));
        assertEquals(-1, pivotInteger(4));
    }

//        public int pivotInteger(int n) {
//        int sum = n * (n + 1) / 2;
//        int leftSum = 0;
//        for (int i = 1; i <= n; i++) {
//            leftSum += i;
//            int rightSum = sum - leftSum + i;
//            if (leftSum == rightSum) return i;
//            if (leftSum > rightSum) break;
//        }
//        return -1;
//    }
//
}
